devito/tools/utils.py

```diff
@@ -156,10 +157,31 @@ def filter_ordered(elements, key=None):
     """
     seen = set()
     if key is None:
-        key = lambda x: x
+        try:
+            unordered, inds = np.unique(elements, return_index=True)
+            return unordered[np.argsort(inds)].tolist()
+        except:
+            return sorted(list(set(elements)), key=elements.index)
     return [e for e in elements if not (key(e) in seen or seen.add(key(e)))]
 
 
+def filter_ordered_index(elements, key=None):
+    """Filter elements in a list while preserving order.
```

Comment:
nitpicking: docstring under quotes

Reply:
the docstring is identical to that of `filter_ordered`

Im not sure I appreciate the difference between this function and `filter_ordered`, on a semantic point of view?